lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -806,15 +569,403 @@ def fetch_gce_metadata(metadata_path)
+    def determine_agent_level_monitored_resource_labels(type)
+      labels = {}
+
+      case type
+
+      # GAE app.
+      when APPENGINE_CONSTANTS[:resource_type]
+        begin
+          labels['module_id'] = fetch_gce_metadata(
+            'instance/attributes/gae_backend_name')
+          labels['version_id'] = fetch_gce_metadata(
+            'instance/attributes/gae_backend_version')
+        rescue StandardError => e
+          @log.error 'Failed to set monitored resource labels for GAE: ',
+                     error: e
+        end
+
+      # GCE.
+      when COMPUTE_CONSTANTS[:resource_type]
+        labels['instance_id'] = @vm_id
+        labels['zone'] = @zone
+
+      # GKE container.
+      when CONTAINER_CONSTANTS[:resource_type]
+        labels['instance_id'] = @vm_id
+        labels['zone'] = @zone
+        begin
+          raw_kube_env = fetch_gce_metadata('instance/attributes/kube-env')
+          kube_env = YAML.load(raw_kube_env)
+          labels['cluster_name'] =
+            cluster_name_from_kube_env(kube_env)
+        rescue StandardError => e
+          @log.error 'Failed to set monitored resource labels for GKE: ',
+                     error: e
+        end
+
+      # Cloud Dataproc.
+      when DATAPROC_CONSTANTS[:resource_type]
+        begin
+          labels['cluster_uuid'] =
+            fetch_gce_metadata('instance/attributes/dataproc-cluster-uuid')
+          labels['cluster_name'] =
+            fetch_gce_metadata('instance/attributes/dataproc-cluster-name')
+          labels['region'] =
+            fetch_gce_metadata('instance/attributes/dataproc-region')
+        rescue StandardError => e
+          @log.error 'Failed to set monitored resource labels for Cloud ' \
+                     'Dataproc: ', error: e
+        end
+
+      # EC2.
+      when EC2_CONSTANTS[:resource_type]
+        labels['instance_id'] = @vm_id
+        labels['region'] = @zone
+        labels['aws_account'] = ec2_metadata['accountId'] if
+          ec2_metadata.key?('accountId')
+      end
+      labels
+    end
+
+    # Determine the common labels that should be added to all log entries
+    # processed by this logging agent.
+    def determine_agent_level_common_labels
+      labels = {}
+      # User can specify labels via config. We want to capture those as well.
+      # TODO: Send instance tags as labels as well?
+      labels.merge!(@labels) if @labels
+
+      case @resource.type
+
+      # GAE app.
+      when APPENGINE_CONSTANTS[:resource_type]
+        labels["#{COMPUTE_CONSTANTS[:service]}/resource_id"] = @vm_id
+        labels["#{COMPUTE_CONSTANTS[:service]}/resource_name"] = @vm_name
+        labels["#{COMPUTE_CONSTANTS[:service]}/zone"] = @zone
+
+      # GCE.
+      when COMPUTE_CONSTANTS[:resource_type]
+        labels["#{COMPUTE_CONSTANTS[:service]}/resource_name"] = @vm_name
+
+      # GKE container.
+      when CONTAINER_CONSTANTS[:resource_type]
```

Comment:
Can GKE and GCE be blended into a single conditional like Dataflow and Dataproc are?